### PR TITLE
PLY Pointcloud

### PR DESCRIPTION
**Pull request by CedricGuillemet**
- moved plys parsing code to loader
- determine if ply is a point cloud, mesh or GS
- added loadAssetContainerAsync
- pointcloud generation
- Mesh generation with vertex color support

image

image

### CODE REVIEW

**Comment by bjsplat** [comment]
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

**Comment by bjsplat** [comment]
Snapshot stored with reference name:
refs/pull/15501/merge

Test environment: 
https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15501/merge/index.html

To test a playground add it to the URL, for example:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15501/merge/index.html#WGZLGJ#4600

Links to test babylon tools with this snapshot:

https://playground.babylonjs.com/?snapshot=refs/pull/15501/merge
https://sandbox.babylonjs.com/?snapshot=refs/pull/15501/merge
https://gui.babylonjs.com/?snapshot=refs/pull/15501/merge
https://nme.babylonjs.com/?snapshot=refs/pull/15501/merge

To test the snapshot in the playground with a playground ID add it after the snapshot query string:

https://playground.babylonjs.com/?snapshot=refs/pull/15501/merge#BCU1XR#0

**Comment by bjsplat** [comment]
Visualization tests for WebGPU (Experimental)

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15501/merge/testResults/webgpuplaywright/index.html

**Comment by bjsplat** [comment]
WebGL2 visualization test reporter:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15501/merge/testResults/webgl2playwright/index.html

File: packages/dev/loaders/src/SPLAT/splatFileLoader.ts [hunk 1]
```diff
@@ -79,49 +103,290 @@ export class SPLATFileLoader implements ISceneLoaderPluginAsync, ISceneLoaderPlu
...
+            };
+        });
+    }
+
+    private _buildPointCloud(pointcloud: PointsCloudSystem, data: ArrayBuffer): boolean {
+        if (!data.byteLength) {
+            return false;
+        }
+        const uBuffer = new Uint8Array(data);
+        const fBuffer = new Float32Array(uBuffer.buffer);
```
Comment (ryantrem) [inline packages/dev/loaders/src/SPLAT/splatFileLoader.ts:132]: Isn't this the same as `new Float32Array(data)`? Assuming yes, that would have been more clear to me.

File: packages/dev/loaders/src/SPLAT/splatFileLoader.ts [hunk 2]
```diff
@@ -79,49 +103,290 @@ export class SPLATFileLoader implements ISceneLoaderPluginAsync, ISceneLoaderPlu
...
+
+        const rowLength = 3 * 4 + 3 * 4 + 4 + 4;
```
Comment (ryantrem) [inline packages/dev/loaders/src/SPLAT/splatFileLoader.ts:134]: Maybe a comment explaining this.

**Review by ryantrem — APPROVED** [review]
Assuming `ConvertPLYToSplat` was just a straight move, so I didn't review it.

File: packages/dev/core/src/Meshes/GaussianSplatting/gaussianSplattingMesh.ts [hunk 3]
```diff
@@ -5,12 +5,12 @@ import { SubMesh } from "../subMesh";
...
 import { Mesh } from "../mesh";
 import { VertexData } from "../mesh.vertexData";
-import { Tools } from "core/Misc/tools";
-import { Matrix, TmpVectors, Vector2, Vector3, Quaternion } from "core/Maths/math.vector";
+import { Matrix, TmpVectors, Vector2, Vector3 } from "core/Maths/math.vector";
 import { Logger } from "core/Misc/logger";
 import { GaussianSplattingMaterial } from "core/Materials/GaussianSplatting/gaussianSplattingMaterial";
 import { RawTexture } from "core/Materials/Textures/rawTexture";
 import { Constants } from "core/Engines/constants";
+import { appendSceneAsync } from "core/Loading/sceneLoader";
```
Comment (sebavan) [inline packages/dev/core/src/Meshes/GaussianSplatting/gaussianSplattingMesh.ts:13]: This is now really strange to have here but I totally get the back compat :-)

**Review by sebavan — APPROVED** [review]

File: packages/dev/core/src/Meshes/GaussianSplatting/gaussianSplattingMesh.ts (hunk 3 above)
Comment (CedricGuillemet) [reply packages/dev/core/src/Meshes/GaussianSplatting/gaussianSplattingMesh.ts:13]: yes, having the load in this class was a mistake. :/

File: packages/dev/loaders/src/SPLAT/splatFileLoader.ts (hunk 2 above)
Comment (CedricGuillemet) [reply packages/dev/loaders/src/SPLAT/splatFileLoader.ts:135]: done

**Comment by sebavan** [comment]
ship it :-)

**Pull request merged**